File: dsa_algo/leetcode/arrays/lc/java/duplicateZeros.java
public class duplicateZeros {
    public static void duplicate_zeros(int[] arr){
        int possible_duplicates = 0;
        int length_ = arr.length - 1;
        for(int l = 0; l <= length_ - possible_duplicates; l++)
        {
            if(arr[l] == 0)
            {
                if(l == length_ - possible_duplicates)
                {
                    arr[length_] = 0;
                    length_ -= 1;
                    break;
                }
                possible_duplicates++;
            }
        }
        int last = length_ - possible_duplicates;
        for(int i = last; i >= 0; i--)
        {
            if(arr[i] == 0)
            {
                arr[i + possible_duplicates] = 0;
                possible_duplicates--;
                arr[i + possible_duplicates] = 0;
            }
            else
            {
                arr[i + possible_duplicates] = arr[i];
            }
        }
    }
}

File: dsa_algo/leetcode/arrays/lc_arrays/python/duplicate_zeros.py
def duplicateZeros(arr: list[int]) -> None:
    possible_duplicates = 0
    length_ = len(arr) - 1
    for left in range(length_ + 1):
        if left > length_ - possible_duplicates:
            break
        if arr[left] == 0:
            if left == length_ - possible_duplicates:
                arr[length_] = 0
                length_ -= 1
                break
            possible_duplicates += 1
    last = length_ - possible_duplicates
    for i in range(last, -1 , -1):
        if arr[i] == 0:
            arr[i + possible_duplicates] = 0
            possible_duplicates -= 1
            arr[i + possible_duplicates] = 0
        else:
            arr[i + possible_duplicates] = arr[i]


arr = [1, 0, 2, 3, 0, 4, 5, 0]
duplicateZeros(arr)
print(arr)